### ISSUE TITLE
Enable `(implicit_transitive_deps false)` only if OCaml version supports `-H`?

### ISSUE THREAD

**Issue by mbarbin** [ocaml/dune#11212]
## Desired Behavior

I recently started using the new `(implicit_transitive_deps false)` with dune `3.17` and OCaml 5 (is `-H` from `5.1` or `5.2`?), and I like it a lot. Thanks!

One issue I did not foresee, and discovered in CI results is that, for certain repos, the CI matrix checks the build with older ocaml versions, such as `4.14`, and because it uses the same dune config files there, things do not work well (this effectively reverts to a world where `-H` is not well supported).

Dune files have the ability to restrict stanza based on the OCaml version, via the `enabled_if` construct. I wonder if something similar could be made available in `dune-project` files.

## Example

Something like this, which I would put in my `dune-project` file maybe?

```lisp
(implicit_transitive_deps (< %{ocaml_version} 5.2))
```
Currently doesn't work:
```
File "dune-project", line 21, characters 1-26:
21 |  (< %{ocaml_version} 5.2))
      ^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Atom expected
```

## Workaround

Maybe I can add a line in the CI script to edit that construct and replace the `false` by `true` for older ocaml version. I haven't tried this yet.

**Comment by mbarbin** [comment]
For context I'm linking an actual repo with my attempt at the workaround approach https://github.com/mbarbin/loc/pull/8

**Comment by nojb** [comment]
Now that OCaml supports -H (since 5.2), there is no longer a reason to support both modes in Dune (other than backwards compatibility), and `(implicit_transitive_deps false)` should become the default (and only) setting of Dune. Is there a reason to support `(implicit_transitive_deps true)` with OCaml >= 5.2?

If we do this, then `(implicit_transitive_deps ...)` could be understood to modify the behaviour only for versions of OCaml < 5.2, which do not support `-H`.

Just an idea.

cc @rgrinberg

**Comment by mbarbin** [comment]
Thanks for commenting @nojb 

> Is there a reason to support `(implicit_transitive_deps true)` with OCaml >= 5.2?

I don't know if some people find it convenient not to have to list transitive dependencies. Imo the new behavior seems desirable and I am happy to forget about the old one. There is though a breaking change to cross when you need to add all the deps that were permitted to be missing, the first time this is enabled. I don't know if this fits with dune policy about breaking changes in general. Perhaps this default can be changed when reaching dune version `4.x`?

**Comment by nojb** [comment]
> let dune decide to completely filter the construct `(implicit_transitive_deps false)` when it knows it is using a version of OCaml that doesn't support it.

This would be a breaking change but that should not cause any code to actually break, but it would mean that older versions of OCaml would no longer be able to use the (partially broken) `(implicit_transitive_deps false)` mode.

An alternative would be to add a third option: `(implicit_transitive_deps only-if-supported)` which would do what you want.

EIther option seems fine for me. I wonder if users would be unhappy if we went with 1... @rgrinberg: do you have an opinion?

**Comment by mbarbin** [comment]
> An alternative would be to add a third option: `(implicit_transitive_deps only-if-supported)` which would do what you want.

I am reading as you meaning `false-only-if-supported`, with `true` being the default in dune.

I didn't think about the fact that this partially supported mode was indeed in use. I like your 3-options proposal. Perhaps we can tweak it so as to hint towards well-supported modes: `true`, `false` (new, mean `false-only-if-supported`), and `false-even-when-partially-supported`, which would be the behavior of the current `false`, for users that really know what they are doing.

I suppose you can call it a breaking change, but depending how you look at it, you can also qualify it a "fix", as I think I had encountered a CI failure for a package that this change would enable to build.

**Comment by nojb** [comment]
> > An alternative would be to add a third option: `(implicit_transitive_deps only-if-supported)` which would do what you want.
> 
> I am reading as you meaning `false-only-if-supported`, with `true` being the default in dune.

Yes, exactly.

> I didn't think about the fact that this partially supported mode was indeed in use.

It was not widely used, but there were a few users (that's how we knew that it was partially broken :)).

> I suppose you can call it a breaking change, but depending how you look at it, you can also qualify it a "fix", as I think I had encountered a CI failure for a package that this change would enable to build.

A (fourth) alternative: make `(implicit_transitive_deps false)` a no-op in OCaml <5.2 but guard the change under a suitable bump of the Dune language version. This would avoid a (technically) breaking change, and would also avoid introducing the third option for `(implicit_transitive_deps)` which is likely to be useless in the long run.

**Comment by mbarbin** [comment]
> make `(implicit_transitive_deps false)` a no-op in OCaml <5.2 but guard the change under a suitable bump of the Dune language version

That sounds good to me.

I don't know about the other few users, I guess if they were experimenting and working towards getting to the now fully supported mode, I can't imagine them being unhappy. Thanks!

**Comment by mbarbin** [comment]
> Perhaps this default can be changed when reaching dune version `4.x`

By the way, if we're still thinking about making `(implicit_transitive_deps false)` the default with dune `4.x`, we'd probably be more comfortable for this to mean `false-if-supported` under the hood, rather than exposing to the users partially supported modes by default. If anything, I take this as an indication that we are on the right track here.

**Comment by rgrinberg** [comment]
We've done this before, but I really dislike changing the meaning of existing options across dune language versions. So I am strongly against changing the meaning of existing values unless such changes are backwards compatible. Moreover, `(implicit_transitive_deps false)` existed, and was in heavy use, years before `-H` was introduced. To call it "partially broken" is kind of a stretch. Especially considering that dune itself has been using it since the beginning.

In summary, any new behavior should be given a value for this field.

**Comment by mbarbin** [comment]
Some of the CI failures I attributed to the combination of `(implicity_transitive_deps false)` and `(ocaml_version < 5.2)` were in fact caused by something unrelated (for example the [linked one](https://github.com/ocaml/opam-repository/pull/27908#issuecomment-2902629722) from yesterday which made me think again about this issue). There is a good chance I am misjudging things here, sorry!

**Comment by nojb** [comment]
> To call it "partially broken" is kind of a stretch.

Yes, I agree.

> In summary, any new behavior should be given a value for this field.

Sounds good. I can prepare such a PR, but it is probably a good idea to wait until there is positive confirmation that this would solve problems out in the wild before doing so. Otherwise, maybe the status quo is enough.

**Comment by mbarbin** [comment]
> Moreover, `(implicit_transitive_deps false)` existed, and was in heavy use, years before `-H` was introduced.

Got it. I suppose I never got to be in that boat, due to some of my dependencies, as reported [here](https://github.com/janestreet/base/issues/161).

If I recall correctly I opened this very dune ticket to try and accommodate projects that do want to enable `(implicit_transitive_deps false)` in their `dune-project`, while using dependencies such as `base` and desire to support both OCaml 4 and 5 builds.

What I ended up doing for now in such projects I own is to fitler out the `(implicit_transitive_deps false)` construct during the CI action workflow when compiling with older compilers. See an example [here](https://github.com/mbarbin/dunolint/blob/cb9f8ab27a910c5d0028eb333ea695774cdc6119/.github/workflows/ocaml-4-ci.yml#L41).

But, now I suspect something changed as I don't think I have encountered build issues in opam-repository' CI for published projects in this combination recently, so I am not sure where things are.

> it is probably a good idea to wait until there is positive confirmation that this would solve problems out in the wild before doing so. Otherwise, maybe the status quo is enough.

👍 Thanks!

**Comment by mbarbin** [comment]
I reproduced a build issue from this CI: https://github.com/ocaml/opam-repository/pull/27912

When compiling with: ocaml-option-flambda.1,ocaml-options-only-flambda.1,ocaml-variants.4.14.2+options

If `(implicit_transitive_deps false)` is set, you get this warning:

```sh
File "_none_", line 1:
(warning 58 [no-cmx-file]): no cmx file was found in path for module Stdune__User_message, and its interface was not compiled with -opaque
```

If  `(implicit_transitive_deps false)` is commented out, the warning doesn't show.

I don't know whether to consider it a limitation of  `(implicit_transitive_deps false)` or something else, but among the options to deal with this, I am thinking that having a way to selectively enable `(implicit_transitive_deps false)` depending on the version of ocaml would be convenient.

Another more extreme take (just a thought) - perhaps I should try and enable `(implicit_transitive_deps false)` only in when in dune dev mode?
